src/drawerbase.js

```diff
@@ -0,0 +1,291 @@
+     */
+    getType(){
+        return undefined;
+    }
+
+    /**
+     * @returns {Boolean} whether the drawer implementation is supported by the browser
+     */
+    static isSupported() {
+        $.console.error('Drawer.isSupported must be implemented by child class');
+    }
+
+    /**
+     * create the HTML element (e.g. canvas, div) that the image will be drawn into
+     * @returns {Element} the element to draw into
+     */
+    _createDrawingElement() {
+        $.console.error('Drawer._createDrawingElement must be implemented by child class');
+        return null;
+    }
+
+    /**
+     * @param {Array} tiledImages - An array of TiledImages that are ready to be drawn
+     */
+    draw(tiledImages) {
+        $.console.error('Drawer.draw must be implemented by child class');
+    }
+
+    /**
+     * @returns {Boolean} True if rotation is supported.
+     */
+    canRotate() {
+        $.console.error('Drawer.canRotate must be implemented by child class');
+    }
+
+    /**
+     * Destroy the drawer (unload current loaded tiles)
+     */
+    destroy() {
+        $.console.error('Drawer.destroy must be implemented by child class');
+    }
+
+    /**
+     * @returns {Boolean} Whether this drawer requires enforcing minimum tile overlap to avoid showing seams.
+     */
+    minimumOverlapRequired() {
+        return false;
+     }
+
+
+    /**
+     * Turns image smoothing on or off for this viewer. Note: Ignored in some (especially older) browsers that do not support this property.
+     *
+     * @function
+     * @param {Boolean} [imageSmoothingEnabled] - Whether or not the image is
+     * drawn smoothly on the canvas; see imageSmoothingEnabled in
+     * {@link OpenSeadragon.Options} for more explanation.
+     */
+    setImageSmoothingEnabled(imageSmoothingEnabled){
+        $.console.error('Drawer.setImageSmoothingEnabled must be implemented by child class');
+    }
+
+    /**
+     * Optional public API to draw a rectangle (e.g. for debugging purposes)
+     * Child classes can override this method if they wish to support this
+     * @param {OpenSeadragon.Rect} rect
+     */
+    drawDebuggingRect(rect) {
+        $.console.warn('[drawer].drawDebuggingRect is not implemented by this drawer');
+    }
+
+    // Deprecated functions
+    clear(){
+        $.console.warn('[drawer].clear() is deprecated. The drawer is responsible for clearing itself as needed before drawing tiles.');
+    }
+
+    // Private functions
+
+    /**
+     * @private
+     * @inner
+     * Ensures that child classes have provided implementations for public API methods
+     * draw, canRotate, destroy, and setImageSmoothinEnabled. Throws an exception if the original
+     * placeholder methods are still in place.
+     */
+    _checkForAPIOverrides(){
+        if(this._createDrawingElement === $.DrawerBase.prototype._createDrawingElement){
+            throw("[drawer]._createDrawingElement must be implemented by child class");
```

Comment:
Looks like we're pretty consistent about always throwing errors instead of just strings. Can you do that here?

Reply:
Done